src/matchbox/match_index.py

```diff
@@ -0,0 +1,218 @@
+# Copyright (C) 2015 by Clearcode <http://clearcode.cc>
+# and associates (see AUTHORS).
+
+# This file is part of matchbox.
+
+# matchbox is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+
+# matchbox is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Lesser General Public License for more details.
+
+# You should have received a copy of the GNU Lesser General Public License
+# along with matchbox.  If not, see <http://www.gnu.org/licenses/>.
+"""Data structure that allows indexing includes and excludes of values."""
+from collections import defaultdict
+
+
+class MatchIndex(object):
+    """
+    An index for matching or mismatching of entities by hashable traits.
+
+    It can answer one question - 'given a trait, what entities are excluded by it?'.
+    It can be used to filter a set of potential matches - not for general querying ('given a trait
+    return all mathing entities').
+    When used as a filter, all entities in the input set must also be indexed by the MatchIndex to be fully filtered.
+
+    .. note ::
+        We can index entities by including or excluding them for given values.
+        matching an object for some values means that this object will match those values and it will NOT
+        match any other values.
+        Mismatching an object for some values means that this object will NOT match those values and will match
+        any other value.
+        It makes no sense to both include and exclude the same object.
+
+    Data layout:
+    We store:
+        - a dict that maps traits to a set of mismatches
+        - a set of entities indexed as matches.
```

Comment:
`indexed as matches`?

Could you elaborate? It's for the `mismatch_unknown` set, right?